plugin/core/promise.py

```diff
@@ -0,0 +1,194 @@
+        """
+        def executor(resolve_fn):
+            return resolve_fn(resolve_value)
+
+        return cls(executor)
+
+    @classmethod
+    def all(cls, promises: List['Promise']) -> 'Promise':
+        """
+        Takes a list of promises and returns a Promise that gets resolved when all promises
+        gets resolved.
+
+        :param      promises: The list of promises
+
+        :returns:   A promise that gets resolved when all passed promises gets resolved.
+                    Gets passed a list with all resolved values.
+        """
+        def handler(resolve):
+            def recheck_resolve_status(_: Any):
+                # We're being called from a Promise that is holding a lock so don't try to use
+                # any methods that would try to acquire it.
+                if all(p.resolved for p in promises):
+                    resolve([p.value for p in promises])
+            for p in promises:
+                if not isinstance(p, Promise):
+                    raise Exception('Value is not a Promise')
+                p.then(recheck_resolve_status)
+
+        if promises:
+            return Promise(handler)
+        return Promise.resolve()
+
+    def __init__(self, executor):
+        """Initialize Promise object.
+
+        Arguments:
+            executor: A function that is executed immediately by this Promise.
+            It gets passed a "resolve" function. The "resolve" function, when
+            called, resolves the Promise with the value passed to it.
+        """
+        self.value = None
+        self.resolved = False
+        self.mutex = threading.Lock()
+        self.callbacks = []
+        self._invoke_executor(executor)
+
+    def __repr__(self):
+        if self.resolved:
+            return 'Promise({})'.format(self.value)
+        return 'Promise(<pending>)'
+
+    def then(self, callback):
+        """Create a new promise and chain it with this promise.
+
+        When this promise gets resolved, the callback will be called with the
+        value that this promise resolved with.
+
+        Returns a new promise that can be used to do further chaining.
+
+        Arguments:
+            callback: The callback to call when this promise gets resolved.
+        """
+        def callback_wrapper(resolve_fn, resolve_value):
+            """A wrapper called when this promise resolves.
+
+            Arguments:
+                resolve_fn: A resolve function of newly created promise.
+                resolve_value: The value with which this promise resolved.
+            """
+            result = callback(resolve_value)
+            # If returned value is a promise then this promise needs to be
+            # resolved with the value of returned promise.
+            if isinstance(result, Promise):
+                result.then(resolve_fn)
+            else:
+                resolve_fn(result)
+
+        def sync_executor(resolve_fn):
```

Comment:
I don't like the wording "executor" here. These functions do not really play the role of an executor (i.e. thread/threadpool/mainthread/eventloop).

Reply:
Changed to `sync_wrapper` / `async_wrapper`